include/irr/builtin/material_compiler/glsl/common.glsl

```diff
@@ -926,57 +1055,193 @@ irr_glsl_BSDFSample irr_bsdf_cos_generate(in instr_stream_t stream, in vec3 rand
+		pdf = 1.0;
+	} else
+#endif
+#if defined(OP_DIFFUSE) || defined(OP_DIFFTRANS)
+	if (op_isDiffuse(op))
+	{
+		vec3 localL = irr_glsl_projected_hemisphere_generate(u.xy);
+		if (is_bsdf)
+		{
+			float dummy;
+			bool flip = irr_glsl_partitionRandVariable(0.5, u.z, dummy);
+			localL = flip ? -localL : localL;
+		}
+		s = irr_glsl_createLightSampleTangentSpace(localV, localL, tangentFrame);
+		out_microfacet = irr_glsl_calcAnisotropicMicrofacetCache(currInteraction, s);
+
+		const float alpha2 = is_bsdf ? 0.0 : ax2;
+		const vec3 reflectance = is_bsdf ? vec3(1.0) : refl;
+		rem *= reflectance*irr_glsl_oren_nayar_cos_remainder_and_pdf(pdf, s, currInteraction.isotropic, alpha2);
+		pdf *= is_bsdf ? 0.5 : 1.0;
 	} else
-#endif //OP_DIFFTRANS
+#endif
+#if defined(OP_CONDUCTOR) || defined(OP_COATING) || defined(OP_DIELECTRIC)
+	if (op_hasSpecular(op))
+	{
+		const float TdotV2 = currInteraction.TdotV * currInteraction.TdotV;
+		const float BdotV2 = currInteraction.BdotV * currInteraction.BdotV;
+		const float NdotV2 = currInteraction.isotropic.NdotV_squared;
+
+		const vec3 upperHemisphereLocalV = currInteraction.isotropic.NdotV < 0.0 ? -localV : localV;
 
-//TODO need to distinguish BSDFs and BRDFs
+		float G2_over_G1 = 0.0;
+		float G1_over_2NdotV = 0.0;
+		float ndf_val = 0.0;
+		vec3 localH = vec3(0.0);
+
+		BEGIN_CASES(ndf)
 #ifdef NDF_GGX
-	if (ndf == NDF_GGX) {
-		s = irr_glsl_ggx_cos_generate(currInteraction, rand.xy, ax, ay);
-	} else
+		CASE_BEGIN(ndf, NDF_GGX) 
+		{
+			// why is it called without "wo_clamps" and beckmann sampling is?
+			localH = irr_glsl_ggx_cos_generate(upperHemisphereLocalV, u.xy, ax, ay);
+		} CASE_END
 #endif //NDF_GGX
 
 #ifdef NDF_BECKMANN
-	if (ndf == NDF_BECKMANN) {
-		s = irr_glsl_beckmann_smith_cos_generate(currInteraction, rand.xy, ax, ay);
-	} else
+		CASE_BEGIN(ndf, NDF_BECKMANN) 
+		{
+			localH = irr_glsl_beckmann_cos_generate_wo_clamps(upperHemisphereLocalV, u.xy, ax, ay);
+		} CASE_END
 #endif //NDF_BECKMANN
 
 #ifdef NDF_PHONG
-	if (ndf == NDF_PHONG) {
-		s = irr_glsl_beckmann_smith_cos_generate(currInteraction, rand.xy, ax, ay);
-	} else
+		CASE_BEGIN(ndf, NDF_PHONG) 
+		{
+			localH = irr_glsl_beckmann_cos_generate_wo_clamps(upperHemisphereLocalV, u.xy, ax, ay);
+		} CASE_END
 #endif //NDF_PHONG
-	{}
+		CASE_OTHERWISE
+		{}
+		END_CASES
+
+		vec3 localL;
+		float VdotH = dot(localH, localV);
+		float VdotH_clamp = irr_glsl_conditionalAbsOrMax(is_bsdf, VdotH, 0.0);
+		vec3 fr;
+#ifdef OP_CONDUCTOR
+		if (op == OP_CONDUCTOR)
+			fr = irr_glsl_fresnel_conductor(ior[0], ior[1], VdotH_clamp);
+		else
+#endif
+			fr = irr_glsl_fresnel_dielectric_common(ior[0]*ior[0], VdotH_clamp);
+		const float refractionProb = dot(CIE_XYZ_Luma_Y_coeffs, fr);
+		float rcpChoiceProb;
+		const bool refraction = is_bsdf ? irr_glsl_partitionRandVariable(refractionProb, u.z, rcpChoiceProb) : false;
+		rcpChoiceProb = is_bsdf ? rcpChoiceProb : 1.0;
+		rem *= rcpChoiceProb;
+		rem *= refraction ? (vec3(1.0) - fr) : fr;
```

Comment:
NOOOOOOOOOOOOOOOOOOOOOOOOOOOOPPPPPPPPPPPPPEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEEE

Conductor ONLY!

Othewise fresnel goes into the PDF!